tools/windows-node-installer/pkg/cloudprovider/aws/ec2.go

```diff
@@ -477,201 +479,178 @@ func (a *AwsProvider) getInfrastructureVPC(infraID string) (*ec2.Vpc, error) {
+
+	ports, hasClusterCidrRule := examineRulesInSg(myIP, rules, vpcCidr)
+
+	// If ClusterCidrRule is missing, add default rule with IpProtocol value "-1"
+	if !hasClusterCidrRule {
+		rulesForUpdate = append(rulesForUpdate,
 			(&ec2.IpPermission{}).
 				// -1 is to allow all ports.
 				SetIpProtocol("-1").
 				SetIpRanges([]*ec2.IpRange{
 					(&ec2.IpRange{}).
-						SetCidrIp(*vpc.CidrBlock),
-				}),
-			(&ec2.IpPermission{}).
-				SetIpProtocol("tcp").
-				SetFromPort(3389).
-				SetToPort(3389).
-				SetIpRanges([]*ec2.IpRange{
-					(&ec2.IpRange{}).
-						SetCidrIp(myIP + "/32"),
-				}),
-			(&ec2.IpPermission{}).
-				SetIpProtocol("tcp").
-				// winrm ansible https port
-				SetFromPort(WINRM_PORT).
-				SetToPort(WINRM_PORT).
-				SetIpRanges([]*ec2.IpRange{
-					(&ec2.IpRange{}).
-						SetCidrIp(myIP + "/32"),
-				}),
+						SetCidrIp(vpcCidr),
+				}))
+	}
+
+	// create a list of rules to be updated in sg
+	rulesForUpdate = append(rulesForUpdate, createRulesFromPorts(ports, myIP)...)
+	return rulesForUpdate
+}
+
+// Create a set of rules to be added in sg given the input ports, helps in creating an appended list of all rules
+func createRulesFromPorts(ports []int64, myIP string) []*ec2.IpPermission {
+	populatedRules := make([]*ec2.IpPermission, 0)
+
+	for _, port := range ports {
+		populatedRules = append(populatedRules,
 			(&ec2.IpPermission{}).
 				SetIpProtocol("tcp").
-				// SSH port to be opened for
-				// TODO: add validation to check if the port is already opened, inform the user
-				// of the same and this should be a no-op
-				SetFromPort(sshPort).
-				SetToPort(sshPort).
+				SetFromPort(port).
+				SetToPort(port).
 				SetIpRanges([]*ec2.IpRange{
 					(&ec2.IpRange{}).
 						SetCidrIp(myIP + "/32"),
-				}),
-		},
-	})
-	if err != nil {
-		return err
+				}))
+		log.Printf(fmt.Sprintf("Added rule with port %d to the security groups of your local IP \n", port))
 	}
-	return nil
+	return populatedRules
 }
 
-// findWindowsWorkerSg finds the Windows worker security group based on security group name <infraID>-windows-worker-sg.
-func (a *AwsProvider) findWindowsWorkerSg(infraID string) (string, error) {
-	sgName := strings.Join([]string{infraID, "windows", "worker", "sg"}, "-")
-	sgs, err := a.EC2.DescribeSecurityGroups(&ec2.DescribeSecurityGroupsInput{
-		Filters: []*ec2.Filter{
-			{
-				Name:   aws.String("group-name"),
-				Values: aws.StringSlice([]string{sgName}),
-			},
-		},
+// examineRulesInSg returns a list of ports which need to be updated in the rules for local IP and a flag indicating
+// whether or not the cluster Id rule with IP protocol "-1" is present in the rule set.
+func examineRulesInSg(myIP string, rules []*ec2.IpPermission, vpcCidr string) ([]int64, bool) {
+	var ports []int64
+	var hasClusterCidrRule bool
+	portTracker := map[int64]bool{}
+
+	// Loop through the rules, then loop through the ips to get a map of ports
+	if rules != nil {
```

Comment:
You can just return if `rules == nil`.  That will make rest of the code easier to read.